File: sdk/python/graphdb_sdk/tenant.py
from __future__ import annotations

from typing import Any
from urllib import parse


class TenantMixin:
    def create_tenant(self, tenant_id: str, **metadata: Any) -> dict:
        return self._json("POST", "/v1/tenants", tenant_id=None, body={"tenant_id": tenant_id, **metadata})

    def list_tenants(self, include_legacy: bool = False) -> dict:
        return self._json("GET", "/v1/tenants", tenant_id=None, query={"include_legacy": include_legacy})

    def get_tenant(self, tenant_id: str) -> dict:
        return self._json("GET", f"/v1/tenants/{self._escape(tenant_id)}", tenant_id=None)

    def disable_tenant(self, tenant_id: str) -> dict:
        return self._tenant_action(tenant_id, "disable")

    def enable_tenant(self, tenant_id: str) -> dict:
        return self._tenant_action(tenant_id, "enable")

    def delete_tenant(self, tenant_id: str) -> dict:
        return self._json("DELETE", f"/v1/tenants/{self._escape(tenant_id)}", tenant_id=None)

    def purge_tenant(self, tenant_id: str, force: bool = False) -> dict:
        return self._json("POST", f"/v1/tenants/{self._escape(tenant_id)}/purge", tenant_id=None, query={"force": force})

    def clone_tenant(self, source_tenant_id: str, target_tenant_id: str, **metadata: Any) -> dict:
        body = {"target_tenant_id": target_tenant_id, **metadata}
        return self._json("POST", f"/v1/tenants/{self._escape(source_tenant_id)}/clone", tenant_id=None, body=body)

    def backup_tenant(self, tenant_id: str) -> dict:
        return self._json("POST", f"/v1/tenants/{self._escape(tenant_id)}/backup", tenant_id=None)

    def restore_tenant(self, tenant_id: str, backup_key: str, overwrite: bool = False, dry_run: bool = False) -> dict:
        body = {"backup_key": backup_key, "overwrite": overwrite, "dry_run": dry_run}
        return self._json("POST", f"/v1/tenants/{self._escape(tenant_id)}/restore", tenant_id=None, body=body)

    def restore_drill_tenant(self, tenant_id: str, params: dict | None = None) -> dict:
        return self._json("POST", f"/v1/tenants/{self._escape(tenant_id)}/restore-drill", tenant_id=None, body=params or {})

    def _tenant_action(self, tenant_id: str, action: str) -> dict:
        return self._json("POST", f"/v1/tenants/{self._escape(tenant_id)}/{action}", tenant_id=None)

    def _escape(self, value: str) -> str:
        return parse.quote(value, safe="")

File: sdk/go/graphdb/tenant.go
package graphdb

import (
	"context"
	"net/url"
	"strconv"
)

type TenantInfo struct {
	TenantID         string            `json:"tenant_id"`
	Status           string            `json:"status"`
	Name             string            `json:"name,omitempty"`
	Description      string            `json:"description,omitempty"`
	Labels           map[string]string `json:"labels,omitempty"`
	Metadata         map[string]any    `json:"metadata,omitempty"`
	ClonedFrom       string            `json:"cloned_from,omitempty"`
	ManifestVersion  int64             `json:"manifest_version"`
	SnapshotVersion  int64             `json:"snapshot_version"`
	CommitTailLength int               `json:"commit_tail_length"`
	Exists           bool              `json:"exists"`
	CreatedAt        string            `json:"created_at,omitempty"`
	UpdatedAt        string            `json:"updated_at,omitempty"`
	DisabledAt       string            `json:"disabled_at,omitempty"`
	DeletedAt        string            `json:"deleted_at,omitempty"`
}

type TenantCreateRequest struct {
	TenantID     string            `json:"tenant_id,omitempty"`
	Name         string            `json:"name,omitempty"`
	Description  string            `json:"description,omitempty"`
	Labels       map[string]string `json:"labels,omitempty"`
	Metadata     map[string]any    `json:"metadata,omitempty"`
	Config       map[string]any    `json:"config,omitempty"`
	SourcePolicy *SourcePolicy     `json:"source_policy,omitempty"`
}

type TenantCloneRequest struct {
	TargetTenantID string            `json:"target_tenant_id"`
	Name           string            `json:"name,omitempty"`
	Description    string            `json:"description,omitempty"`
	Labels         map[string]string `json:"labels,omitempty"`
	Metadata       map[string]any    `json:"metadata,omitempty"`
}

type TenantPurgeReport struct {
	TenantID             string   `json:"tenant_id"`
	Deleted              int      `json:"deleted"`
	DeletedKeys          []string `json:"deleted_keys,omitempty"`
	DeletedKeysTruncated bool     `json:"deleted_keys_truncated,omitempty"`
}

func (c *Client) ListTenants(ctx context.Context, includeLegacy bool) ([]TenantInfo, error) {
	values := url.Values{}
	setBool(values, "include_legacy", includeLegacy)
	var out struct {
		Tenants []TenantInfo `json:"tenants"`
	}
	if err := c.doJSON(ctx, "GET", "/v1/tenants", "", values, nil, &out); err != nil {
		return nil, err
	}
	return out.Tenants, nil
}

func (c *Client) CreateTenant(ctx context.Context, request TenantCreateRequest) (out TenantInfo, err error) {
	err = c.doJSON(ctx, "POST", "/v1/tenants", "", nil, request, &out)
	return out, err
}

func (c *Client) GetTenant(ctx context.Context, tenantID string) (out TenantInfo, err error) {
	err = c.doJSON(ctx, "GET", "/v1/tenants/"+pathEscape(tenantID), "", nil, nil, &out)
	return out, err
}

func (c *Client) UpdateTenant(ctx context.Context, tenantID string, request TenantCreateRequest) (out TenantInfo, err error) {
	err = c.doJSON(ctx, "PUT", "/v1/tenants/"+pathEscape(tenantID), "", nil, request, &out)
	return out, err
}

func (c *Client) DisableTenant(ctx context.Context, tenantID string) (TenantInfo, error) {
	return c.tenantAction(ctx, tenantID, "disable")
}

func (c *Client) EnableTenant(ctx context.Context, tenantID string) (TenantInfo, error) {
	return c.tenantAction(ctx, tenantID, "enable")
}

func (c *Client) DeleteTenant(ctx context.Context, tenantID string) (out TenantInfo, err error) {
	err = c.doJSON(ctx, "DELETE", "/v1/tenants/"+pathEscape(tenantID), "", nil, nil, &out)
	return out, err
}

func (c *Client) PurgeTenant(ctx context.Context, tenantID string, force bool) (out TenantPurgeReport, err error) {
	values := url.Values{}
	setBool(values, "force", force)
	err = c.doJSON(ctx, "POST", "/v1/tenants/"+pathEscape(tenantID)+"/purge", "", values, nil, &out)
	return out, err
}

func (c *Client) CloneTenant(ctx context.Context, sourceTenantID string, request TenantCloneRequest) (out TenantInfo, err error) {
	err = c.doJSON(ctx, "POST", "/v1/tenants/"+pathEscape(sourceTenantID)+"/clone", "", nil, request, &out)
	return out, err
}

func (c *Client) BackupTenant(ctx context.Context, tenantID string) (out Task, err error) {
	err = c.doJSON(ctx, "POST", "/v1/tenants/"+pathEscape(tenantID)+"/backup", "", nil, nil, &out)
	return out, err
}

// BackupTenantToObjectStorage captures a committed snapshot in the server's
// configured S3 repository. The completed task result contains its backup_key.
func (c *Client) BackupTenantToObjectStorage(ctx context.Context, tenantID string) (out Task, err error) {
	err = c.doJSON(ctx, "POST", "/v1/tenants/"+pathEscape(tenantID)+"/backup", "", nil, map[string]any{"destination": "object"}, &out)
	return out, err
}

type ObjectBackup struct {
	Format      string `json:"format"`
	TenantID    string `json:"tenant_id"`
	BackupID    string `json:"backup_id"`
	Version     int64  `json:"version"`
	CreatedAt   string `json:"created_at"`
	SnapshotKey string `json:"snapshot_key"`
	Bytes       int64  `json:"bytes"`
	SHA256      string `json:"sha256"`
	BackupKey   string `json:"backup_key"`
}

type ObjectBackupPage struct {
	Backups    []ObjectBackup `json:"backups"`
	NextCursor string         `json:"next_cursor,omitempty"`
}

func (c *Client) ListObjectBackups(ctx context.Context, tenantID, cursor string, limit int) (out ObjectBackupPage, err error) {
	values := url.Values{}
	if cursor != "" {
		values.Set("cursor", cursor)
	}
	if limit != 0 {
		values.Set("limit", strconv.Itoa(limit))
	}
	err = c.doJSON(ctx, "GET", "/v1/tenants/"+pathEscape(tenantID)+"/backups", "", values, nil, &out)
	return out, err
}

func (c *Client) RestoreTenant(ctx context.Context, tenantID string, backupKey string, overwrite bool, dryRun bool) (out Task, err error) {
	request := map[string]any{"backup_key": backupKey, "overwrite": overwrite, "dry_run": dryRun}
	err = c.doJSON(ctx, "POST", "/v1/tenants/"+pathEscape(tenantID)+"/restore", "", nil, request, &out)
	return out, err
}

func (c *Client) RestoreDrillTenant(ctx context.Context, tenantID string, request map[string]any) (out Task, err error) {
	err = c.doJSON(ctx, "POST", "/v1/tenants/"+pathEscape(tenantID)+"/restore-drill", "", nil, request, &out)
	return out, err
}

func (c *Client) tenantAction(ctx context.Context, tenantID string, action string) (out TenantInfo, err error) {
	err = c.doJSON(ctx, "POST", "/v1/tenants/"+pathEscape(tenantID)+"/"+action, "", nil, nil, &out)
	return out, err
}
